### ISSUE TITLE
fast memory leak with broken?/misbehaving web seed

### ISSUE THREAD

**Issue by 2L3R4** [transmission/transmission#4595]
### What is the issue?

TLDR: transmission daemon allocates memory at download speed while downloading from broken/misbehaving web seed, then gets killed by OOM killer of the kernel

After adding the torrent file for the latest manjaro-sway ( https://mirror.manjaro-sway.download/manjaro-sway/manjaro-sway.torrent ): [manjaro-sway.torrent.txt](https://github.com/transmission/transmission/files/10419897/manjaro-sway.torrent.txt) to the webinterface, transmission-daemon is trying to download at the full speed (if no download speed limit is set).

But the progress meter stops (in the current release of manjaro-sway) at ~50MB, while it further downloads at ~90MB/s (if internet connection is fast enough). I assume that these 90MB/s gets allocated as RAM until the system runs out.

if one sets the download limit to something like 5MB/s the RAM gets slowly filled up.

crash log:

[transmission-daemon-crash.log](https://github.com/transmission/transmission/files/10419911/transmission-daemon-crash.log)

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

3.00 (bb6b5a062e) on debian 11

**Comment by ckerr** [comment]
Hi @2L3R4, thanks for reporting this! Please give the Transmission 4 series a try.

Duplicate of https://github.com/transmission/transmission/issues/3494

**Comment by boredland** [comment]
I don't really get a much better result with v4. It runs at apparent 20Mb/s while not adding much to the downloaded artifact (after minutes it is still at ~50Mb).

![image](https://user-images.githubusercontent.com/4662748/212565279-e4623949-0414-46b0-84b0-fe72e312c0e9.png)

**Comment by 2L3R4** [comment]
> I don't really get a much better result with v4. It runs at apparent 20Mb/s while not adding much to the downloaded artifact (after minutes it is still at ~50Mb).
> 
> ![image](https://user-images.githubusercontent.com/4662748/212565279-e4623949-0414-46b0-84b0-fe72e312c0e9.png)
> 

does v4 still (try to) allocate like many hundert MBs of RAM?

**Comment by boredland** [comment]
it starts at around 860Mb and fastly konverges to 2Gb where it stalls.

**Comment by ckerr** [comment]
@2L3R4 you're right, this is a special case related to misbehaving webseeds.

The upstream issue is that https://mirror.manjaro-sway.download is not responding correctly to [HTTP range requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests). I can reproduce this even outside of Transmission, e.g.

```sh
$ https://mirror.manjaro-sway.download/manjaro-sway/manjaro-sway-21.3.2-230115-linux515.iso
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
 49 2106M   49 1045M    0     0  47.0M      0  0:00:44  0:00:22  0:00:22 53.3M^C
```

will just keep downloading indefinitely instead of returning 101 bytes from [100-200].

The problem _in Transmission_ is that it does pretty much what curl is doing here; it keeps downloading even though the size exceeds the range that requested.

So probably Transmission should bail out once the buffered DL size exceeds the requested range.

**Comment by ckerr** [comment]
Two different calls to `curl --range "100-200" -D headers2.txt --output foo https://mirror.manjaro-sway.download/manjaro-sway/manjaro-sway-21.3.2-230115-linux515.iso` give different headers:

```
HTTP/1.1 200 OK
Date: Sun, 15 Jan 2023 21:45:44 GMT
Content-Length: 2208436224
Connection: keep-alive
```

and

```
HTTP/1.1 206 Partial Content
Date: Sun, 15 Jan 2023 21:45:58 GMT
Content-Length: 101
Connection: keep-alive
Content-Range: bytes 100-200/2208436224
```

So it looks like _some_ over the servers in the pool work, while others don't.

Anyway, this gives us two easy tests that we can use to bail out:

 - If we requested a range but get back a 200 header instead of 206
 - If we get a Content-Length that doesn't match the expected size

**Comment by boredland** [comment]
That's super interesting. For the record: it is just a cloudflare [r2](https://www.cloudflare.com/products/r2/) bucket (their take at s3) set to "public access":

![image](https://user-images.githubusercontent.com/4662748/212569229-77cbafa0-4560-4a64-9286-52bd9935555c.png)

**Comment by ckerr** [comment]
I've added code that stops using webseed servers that don't honor partial content requests. This WfM when testing on the Manjaro torrent, but since there aren't many peers this is slower than if there was a working webseed. You may want to report this issue upstream to Manjaro as well so that they fix their server.

If you'd like to test out the Transmission fix, it will be in the [nightly builds](https://transmissionbt.com/download).

**Comment by boredland** [comment]
as said - we don't 'host' those files, but I will try to forward this to sb at cloudflare!

**Comment by 2L3R4** [comment]
"manjaro-sway" is from what I have heard a project by @boredland (<- person with most commits in the manjaro-sway repos, so it may be hosted on personal infrastructure) except if the method of creating the webseed is a issue created by upstream manjaro repos (which I doubt)

**Comment by ckerr** [comment]
@boredland ah! I didn't understand that you were here representing Manjaro. Now my comment about reporting this upstream doesn't make much sense :smile:
